cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.provision;
+
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.provisioner.spi.Provisioner;
+import co.cask.cdap.provisioner.spi.ProvisionerSpec;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * Service for provisioning related operations
+ */
+public class ProvisioningService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(ProvisioningService.class);
+  private final ProvisionerExtensionLoader provisionerExtensionLoader;
+  private final AtomicReference<ProvisionerInfo> provisionerInfo;
+
+  @Inject
+  ProvisioningService(CConfiguration cConf) {
+    provisionerExtensionLoader = new ProvisionerExtensionLoader(cConf.get(Constants.Provisioner.EXTENSIONS_DIR));
+    provisionerInfo = new AtomicReference<>(new ProvisionerInfo(new HashMap<>(), new HashMap<>()));
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    reloadProvisioners();
```

Comment:
Will this throw exception? If not, can it be done in the constructor instead? The reason is, it seems like this class doesn't really looks like a Service (i.e. something that has a start/stop lifecycle), and making it a Service meaning it involves explicit start/stop call before using it.

Reply:
I think the only exception it can throw is SecurityExceptions if the extension directory is not readable. I suppose that is something we can verify at startup though and make this not a Service.